### PR TITLE
Propagate var binds across list comprehension filters

### LINKED ISSUE #442: badkey crash in typechecker:do_type_check_expr_in/3
Using latest master (089ac86a88570509dcfe86d2189fb0fb27b9ff35), I get this crash:
```
Error: {badkey,'Id'}
[{erlang,map_get,
         ['Id',
          #{'AllFun' => {type,0,'fun',[{type,0,any},{type,0,any,[]}]},
            'Item' => {type,0,any,[]},
            'RunningItems' => {type,0,list,[{user_type,0,subscriber_info,[]}]},
            'SourceId' => {type,0,atom,[]},
            'UpdateFun' => {type,0,'fun',[{type,0,any},{type,0,any,[]}]},
            'UpdateRunning' => {type,0,union,[{atom,0,false},{atom,0,true}]}}],
         []},
 {typechecker,do_type_check_expr_in,3,
              [{file,"/home/jespereskilson/dev/kred/_build/default/plugins/gradualizer/src/typechecker.erl"},
               {line,2337}]},
 {typechecker,type_check_expr_in,3,
              [{file,"/home/jespereskilson/dev/kred/_build/default/plugins/gradualizer/src/typechecker.erl"},
               {line,2325}]},
 {typechecker,'-type_check_call_ty/4-lc$^0/1-0-',2,
              [{file,"/home/jespereskilson/dev/kred/_build/default/plugins/gradualizer/src/typechecker.erl"},
               {line,2134}]},
 {typechecker,type_check_call_ty,4,
              [{file,"/home/jespereskilson/dev/kred/_build/default/plugins/gradualizer/src/typechecker.erl"},
               {line,2135}]},
 {typechecker,do_type_check_expr,2,
              [{file,"/home/jespereskilson/dev/kred/_build/default/plugins/gradualizer/src/typechecker.erl"},
               {line,1677}]},
 {typechecker,type_check_expr,2,
              [{file,"/home/jespereskilson/dev/kred/_build/default/plugins/gradualizer/src/typechecker.erl"},
               {line,1554}]},
 {typechecker,type_check_rel_op_in,3,
              [{file,"/home/jespereskilson/dev/kred/_build/default/plugins/gradualizer/src/typechecker.erl"},
               {line,2926}]}]
```
The function being analyzed is (I think):
```erlang
-spec update_users_of_source(SourceId :: kafka_kred:source(),
                             UpdateRunning :: boolean(),
                             UpdateFun :: fun(),
                             CollectorFun :: fun(),
                             RunningItems :: [subscriber_info()]) ->
                              list(ok | {error, Reason :: any()}).
update_users_of_source(SourceId, UpdateRunning, UpdateFun,
                       AllFun, RunningItems) ->
  [UpdateFun(Item)
   || Item <- AllFun(),
   begin
     #subscriber_info{id = Id, source = Source} = Item,
     Source =:= SourceId
   end,
   UpdateRunning =:= lists:keymember(Id, 1, RunningItems)
  ].
```

### PR DESCRIPTION
**Pull request by erszcz**
This fixes #442.

### CODE REVIEW

File: src/typechecker.erl [hunk 1]
```diff
@@ -3129,8 +3129,8 @@ type_check_comprehension_in(Env, ResTy, OrigExpr, Compr, Expr, P, [Pred | Quals]
     %% TODO: As a non-boolean predicates don't give runtime errors, we could
     %% possibly add a configuration parameter to toggle this check.
     {VB1, Cs1} = type_check_expr_in(Env, {type, erl_anno:new(0), 'boolean', []}, Pred),
-    {VB2, Cs2} = type_check_comprehension_in(Env, ResTy, OrigExpr, Compr, Expr, P, Quals),
-    {union_var_binds(VB1, VB2, Env), constraints:combine(Cs1, Cs2)}.
+    {VB2, Cs2} = type_check_comprehension_in(union_var_binds([VB1], Env), ResTy, OrigExpr, Compr, Expr, P, Quals),
+    {union_var_binds([VB2], Env), constraints:combine(Cs1, Cs2)}.
```
Comment (zuiderkwast) [inline src/typechecker.erl:3133]: Shouldn't VB1 be returned too? I'm thinking about a comprehension like `[Id || begin Id = 1, true end]`.

```suggestion
    Env1 = union_var_binds([VB1], Env),
    {VB2, Cs2} = type_check_comprehension_in(Env1, ResTy, OrigExpr, Compr, Expr, P, Quals),
    {union_var_binds([VB2], Env1), constraints:combine(Cs1, Cs2)}.
```

Comment (erszcz) [reply src/typechecker.erl]: It was already folded into `VB2`. The following passed on `f` and, as expected, failed on `g`:

```
-spec f() -> [integer()].
f() ->
    [Id || begin Id = 1, true end].

-spec g() -> [atom()].
g() ->
    [Id || begin Id = 1, true end].
```

However, your suggestion made it more explicit, so let's keep it :+1:

Comment (zuiderkwast) [reply src/typechecker.erl]: > It was already folded into VB2.

Are you sure? I though the returned VB2 only contains new var bindings from the remaining generators and filters (Quals). I may be wrong though.

Comment (erszcz) [reply src/typechecker.erl]: This commit - https://github.com/josefs/Gradualizer/pull/444/commits/dd037e5eafad8d2e6b62d415df3a789f438b8d09 - folds `VB1` into `Env` before passing it down to `type_check_comprehension_in`. When we get `VB2` from it, `VB1` is already included.

**Pull request merged**